### PR TITLE
Timing fixes

### LINKED ISSUE #563: What's the difference among many timeouts
There're many timeout related configurations, but some of those comments are a little bit vague.

```
	// Maximum duration for reading the full request (including body).
	//
	// This also limits the maximum duration for idle keep-alive
	// connections.
	//
	// By default request read timeout is unlimited.
	ReadTimeout time.Duration

	// Maximum duration for writing the full response (including body).
	//
	// By default response write timeout is unlimited.
	WriteTimeout time.Duration
```

* For ReadTimeout, when it starts timing? Does it mean that a keep-alive connection can be used no more than ReadTimeout, no matter how many times it has been used? What's the difference between ReadTimeout and MaxKeepaliveDuration?
* WriteTimeout, when it starts timing?
* What should I set if I expect an incoming request is processed no more than X seconds,  and will get a error if timeout?

### PR DESCRIPTION
**Pull request by erikdubbelboer**
- Improved timing code to be much cleaner
- Add IdleTimeout
- Remove obsolete optimization (was fixed in Go 2 years ago: https://github.com/golang/go/issues/15133#issuecomment-271571395)

Fixes #563 and potentially #491

### CODE REVIEW

**Comment by erikdubbelboer** [comment]
@kirillDanshin what do you think about this? I have been testing this in a high RPS environment for some days now and it performs really well.

File: server.go [hunk 1]
```diff
@@ -199,19 +199,26 @@ type Server struct {
 	// Default buffer size is used if not set.
 	WriteBufferSize int
 
-	// Maximum duration for reading the full request (including body).
-	//
-	// This also limits the maximum duration for idle keep-alive
-	// connections.
+	// ReadHeaderTimeout is the amount of time allowed to read
```
Comment (kirillDanshin) [inline server.go]: ReadTimeout

File: server.go [hunk 2]
```diff
@@ -226,15 +233,9 @@ type Server struct {
...
-	// Maximum keep-alive connection lifetime.
-	//
-	// The server closes keep-alive connection after its' lifetime
-	// expiration.
-	//
-	// See also ReadTimeout for limiting the duration of idle keep-alive
-	// connections.
-	//
-	// By default keep-alive connection lifetime is unlimited.
+	// Deprecated!
```
Comment (kirillDanshin) [inline server.go]: https://rakyll.org/deprecated/

File: server.go (hunk 1 above)
Comment (erikdubbelboer) [reply server.go]: Good catch.

File: server.go (hunk 2 above)
Comment (erikdubbelboer) [reply server.go]: I didn't know that!

**Comment by erikdubbelboer** [comment]
@kirillDanshin any thoughts on this? Otherwise I'll merge it as it seems to be good now.

**Pull request merged**